SCMOD-7698: Fix workflow poison msg handling

rorytorneymf opened the pull request:
When the workflow-worker is processing a document, if the document
already has values for the CAF_WORKFLOW_SETTINGS field, it means
that the workflow worker has already resolved the arguments and added
them to the document.

This scenario can occur whenever a downstream worker sends a poison
message back to the workflow worker.

In this case, the ArgumentsManager should not try to re-resolve the arguments
again, but instead just copy the value from CAF_WORKFLOW_SETTINGS from
the document field  into the custom data of the document task response, and
return.

Validation has also been added here to guard against unexpected settings
present in CAF_WORKFLOW_SETTINGS, but not present in the workflow
arguments.

Full description of how to reproduce and examples of poison messages
can be found in the Jira.

rorytorneymf committed:
SCMOD-7698: Fixing workflow-worker poison msg handling.

When the workflow-worker is processing a document, if the document
already has values for the CAF_WORKFLOW_SETTINGS field, it means
that the workflow worker has already resolved the arguments and added
them to the document.

This scenario can occur whenever a downstream worker sends a poison
message back to the workflow worker.

In this case, the ArgumentsManager should not try to re-resolve the arguments
again, but instead just copy the value from CAF_WORKFLOW_SETTINGS from
the document field  into the custom data of the document task response, and
return.

Validation has also been added here to guard against unexpected settings
present in CAF_WORKFLOW_SETTINGS, but not present in the workflow
arguments.

Full description of how to reproduce and examples of poison messages
can be found in the Jira.

andyreidz commented on worker-workflow/src/main/java/com/github/cafdataprocessing/workflow/WorkflowWorker.java:
```diff
@@ -129,8 +130,14 @@ public void processDocument(final Document document) throws DocumentWorkerTransi
+            LOG.error(String.format("UnexpectedCafWorkflowSettingException for document [%s].\n%s\n", 
+                                    document.getReference(), e.toString()));
+            document.addFailure("WORKFLOW_UNEXPECTED_CAF_WORKFLOW_SETTING_EXCEPTION", e.getMessage());
+        }
```
return here, no point in continuing.

rorytorneymf replied:
Ah yes, thanks, missed this.

rorytorneymf committed:
SCMOD-7698: Adding missing return statement.

rorytorneymf committed:
SCMOD:7698 Compile failure temp workaround (DO NOT MERGE)

dermot-hardy reviewed (changes requested):
I believe this fix introduces a security issue.  What is to prevent a rogue agent from staging documents that already contain `CAF_WORKFLOW_SETTINGS` fields?  Properly authored, could that field not be used to write to, or delete from, another tenant's index?!  The tenant id passed in custom data cannot be controlled by a rogue agent and is therefore trustworthy but this change is using this untrustworthy field value in preference to it.

rorytorneymf committed:
SCMOD-7698: Addressing security issue.

Adding an additonal check to ensure we're really dealing
with a poison message (and not a msg staged by a rogue agent)
before accepting that the CAF_WORKFLOW_SETTINGS on the
document are valid.

rorytorneymf committed:
SCMOD-7698: Minor renaming.

dermot-hardy commented on worker-workflow-testing/src/main/java/com/github/cafdataprocessing/workflow/testing/models/ApplicationMock.java:
```diff
@@ -61,6 +61,24 @@ public Application getApplication()
+    public String getVersion()
+    {
+        return "";
+    }
```
#58 has been merged so you should be able to get rid of this now.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/workflow/ArgumentsManager.java:
```diff
@@ -177,5 +198,51 @@ public void checkHealth() {
+        // a rogue agent, and as such we can safely use the CAF_WORKFLOW_SETTINGS present
+        // in the document and trust that they are valid.
+        return document.getField("CAF_WORKFLOW_SETTINGS").hasValues() && 
+            (document.getTask().getCustomData("tenantId") == null);
```
As discussed, let's change this to look for "workflowName" rather than "tenantId", given that tenant is not a core concept to the worker.

dermot-hardy commented on worker-workflow/src/main/java/com/github/cafdataprocessing/workflow/ArgumentsManager.java:
```diff
@@ -177,5 +198,51 @@ public void checkHealth() {
+            (document.getTask().getCustomData("tenantId") == null);
+    }
 
+    private void validateExistingCafWorkflowSettings(final String cafWorkflowSettingsJson,
```
As discussed let's scrap this as not required.

rorytorneymf committed:
Merge branch 'develop' of github.com:CAFDataProcessing/worker-workflow into scmod-7698

rorytorneymf committed:
SCMOD-7698: Updating poison message checks.

To check if a message/document is poison we'll now
check that:

1. It has a a non-empty CAF_WORKFLOW_SETTINGS field.

2. It does NOT have a 'workflowName' inside it's custom data.

Previously we were checking for the absense of a tenantId inside
the custom data, but there are cases where a tenantId may not be
provided (I don't think we have any concrete cases of this at
the moment, but it is allowable).

Instead we'll now check for the absense of a workflowName inside
custom data, there are currently no scenarios where a valid
document will not have this, so if we don't have it, we have some
more confidence we are dealing with a poison message.

Also removed the validation related to CAF_WORKFLOW_SETTINGS
as we didn't feel it was adding much value.

rorytorneymf committed:
SCMOD-7698: Fixing comment typo.

rorytorneymf committed:
SCMOD-7698: Minor cleanup.

dermot-hardy committed:
Removed some whitespace changes
